### PR TITLE
Add Known Origin to nft.events

### PR DESCRIPTION
**Pull request by catherine-2**
Brief comments on the purpose of your changes:

**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by catherine-2** [comment]
@jeff-dude  This one should be ready for review but getting some errors unrelated to the changes in the PR. Both tests were passing until I added the known origin files to nft.events, nft.trades and nft.fees. I'm not sure what is causing these or how to fix

Error in "pre"-commit hooks:
- Missing model (ref) model

Error in slim ci:
- Failure in test opensea_ethereum_assert_samples (tests/opensea/ethereum/opensea_ethereum_assert_samples.sql) Got 1 result, configured to fail if != 0

**Comment by 0xRobin** [comment]
I'll have a look at this.
The pre-commit error is known and can be safely ignored.
The `opensea_ethereum_assert_samples.sql` test was probably not triggered on the opensea v1 rewrite, and might contain false positives.

**Comment by catherine-2** [comment]
> I'll have a look at this.
> The pre-commit error is known and can be safely ignored.
> The `opensea_ethereum_assert_samples.sql` test was probably not triggered on the opensea v1 rewrite (it references nft.trades and not opensea.trades specifically), and might contain false positives.

Awesome thank you @0xRobin !

**Comment by catherine-2** [comment]
@0xRobin  thank you for fixing that issue earlier today! I merged the latest master and both checks are now passing. I checked the test schema table on Dune and everything looks good but let me know if you see any issues. Thank you!

File: models/knownorigin/knownorigin_ethereum_events.sql [hunk 1]
```diff
@@ -0,0 +1,125 @@
+{{ config(
```
Comment (jeff-dude) [inline models/knownorigin/knownorigin_ethereum_events.sql:1]: minor ask: can you nest these in an `ethereum` directory within `knownorigin`?

File: models/knownorigin/knownorigin_ethereum_events.sql [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            , _buyer as buyer
+            , _price as raw_amount
+            , a.evt_block_number
+            , a.evt_index
+            , a.evt_tx_hash
+            , case when b.from = '0x0000000000000000000000000000000000000000' then 'Primary' else 'Secondary' end as sale_type
+            , 'Buy' as trade_category
+            , lower('0xc322cdd03f34b6d25633c2abbc8716a058c7fe9e') as project_contract_address
+    from {{ source('knownorigin_ethereum','TokenMarketplaceV2_evt_TokenPurchased') }} a
+    inner join {{ source('erc721_ethereum','evt_transfer') }} b on a.evt_tx_hash = b.evt_tx_hash and a._tokenId = b.tokenId
```
Comment (jeff-dude) [inline models/knownorigin/knownorigin_ethereum_events.sql:29]: add incremental filter on `evt_transfer` within join conditions, similar to prior PR we did, plz 🤝

Comment (jeff-dude) [reply models/knownorigin/knownorigin_ethereum_events.sql:29]: same idea throughout

File: models/knownorigin/knownorigin_ethereum_events.sql [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        , cast('2.5' as double) as platform_fee_percentage
+        , round((12.5 * (raw_amount) / 100), 7) as royalty_fee_amount_raw
+        , round((12.5 * (raw_amount/1e18) / 100), 7) as royalty_fee_amount
+        , round((12.5 * (raw_amount/1e18) * ep.price / 100), 7) as royalty_fee_amount_usd
+        , cast('12.5' as double) as royalty_fee_percentage
+        , cast(null as varchar(5)) as royalty_fee_receive_address
+        , 'ETH' as royalty_fee_currency_symbol
+        , 'knownorigin' || '-' || a.evt_tx_hash || '-' || a.token_id || '-' ||  seller || '-' || a.evt_block_number || '-' || a.evt_index as unique_trade_id
+from (select * from raw_ko_transactions where sale_type = 'Secondary') a
+inner join {{ source('ethereum','transactions') }} t on a.evt_tx_hash = t.hash
```
Comment (jeff-dude) [inline models/knownorigin/knownorigin_ethereum_events.sql:115]: filter on `transactions` join condition as well

File: models/knownorigin/knownorigin_ethereum_events.sql [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            , lower('0xfbeef911dc5821886e1dda71586d90ed28174b7d') as project_contract_address    
+    from {{ source('knownorigin_ethereum','KnownOriginDigitalAssetV2_evt_Purchase') }} a
+    inner join {{ source('erc721_ethereum','evt_transfer') }} b on a.evt_tx_hash = b.evt_tx_hash and a._tokenId = b.tokenId
+    {% if is_incremental() %}
+    where a.evt_block_time >= date_trunc("day", now() - interval '1 week')
+    {% endif %}
+
+)
+
+select  'ethereum' as blockchain
```
Comment (jeff-dude) [inline models/knownorigin/knownorigin_ethereum_events.sql:76]: can you explicitly cast columns that are done so [here](https://github.com/duneanalytics/spellbook/blob/main/models/archipelago/ethereum/archipelago_ethereum_events.sql)?

Comment (jeff-dude) [reply models/knownorigin/knownorigin_ethereum_events.sql:76]: we want to be consistent across nft sector event tables

File: models/knownorigin/knownorigin_ethereum_fees.sql [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+SELECT blockchain,
+project,
+version,
+block_time,
+token_id,
+collection,
+platform_fee_amount_raw,
+platform_fee_amount,
+platform_fee_amount_usd,
+CAST(platform_fee_percentage AS DOUBLE) AS platform_fee_percentage,
```
Comment (jeff-dude) [inline models/knownorigin/knownorigin_ethereum_fees.sql:19]: if these get cast at the events table, then the cast can likely be removed here. i know a few views on other projects still have the cast, but as long as event table has it, we'll be covered as the view pulls from underlying table

File: models/knownorigin/knownorigin_ethereum_trades.sql [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+amount_usd,
+token_standard,
+trade_type,
+number_of_items,
+trade_category,
+evt_type,
+seller,
+buyer,
+amount_original,
+CAST(amount_raw AS DECIMAL(38,0)) AS amount_raw,
```
Comment (jeff-dude) [inline models/knownorigin/knownorigin_ethereum_trades.sql:25]: same idea on cast here, likely not needed if done at events level

**Comment by jeff-dude** [comment]
@0xRobin can help you finalize with above comments. i'll likely be off on holiday, so he can work with another teammate on helping merge when the time comes 👍

**Comment by catherine-2** [comment]
> 

Thank you! Just addressed all of your comments in my latest commit. Hope you enjoy time off over the holidays!

@0xRobin Whenever you have a chance to take a look would appreciate your thumbs up as well and approval to merge. Thank you!

**Comment by 0xRobin** [comment]
@catherine-2 as discussed on discord, I just pushed some changes (mostly restructuring of the query) and some minor fixes. I've replaced your 'primary/secondary' sale_type to be reflected as 'mint/trade' in evt_type 

We're still missing the other events as you pointed to https://dune.com/queries/1335015/2292301
Once we get all the correct event sources in here we should also:
- make a test seed with samples, preferably 1 or more from every event source / marketplace version, I can help with the setup here. 
- make knownorigin.mints and integrate it into nft.mints

**Comment by 0xRobin** [comment]
Hey @catherine-2, are you still working on this?
If not, I can take this over if you don't mind that.

**Comment by catherine-2** [comment]
> Hey @catherine-2, are you still working on this?
> If not, I can take this over if you don't mind that.

Apologies for the delayed response! Will make updates in the next few days to address the changes we discussed last + conflicts. Will tag you once ready for re-review!

**Comment by jeff-dude** [comment]
closing due to inactivity

**Pull request closed without merging**